File: core/pkg/protocol/outcome_health.go
package protocol

import "strings"

// OutcomeHealthState is the user-facing operational state for an Outcome.
// Stored runtime statuses remain source-specific; this projection gives UI and
// API clients one stable trust/recovery vocabulary.
type OutcomeHealthState string

const (
	OutcomeHealthHealthy   OutcomeHealthState = "healthy"
	OutcomeHealthWaiting   OutcomeHealthState = "waiting"
	OutcomeHealthRunning   OutcomeHealthState = "running"
	OutcomeHealthDegraded  OutcomeHealthState = "degraded"
	OutcomeHealthBlocked   OutcomeHealthState = "blocked"
	OutcomeHealthCompleted OutcomeHealthState = "completed"
	OutcomeHealthArchived  OutcomeHealthState = "archived"
)

func OutcomeHealthForProject(item OutcomeProject) OutcomeHealthState {
	if item.Status == OutcomeProjectStatusArchived {
		return OutcomeHealthArchived
	}
	if len(item.RecoveryRefs) > 0 || item.Status == OutcomeProjectStatusNeedsAttention {
		return OutcomeHealthDegraded
	}
	if item.Status == OutcomeProjectStatusOutputReady || len(item.OutputRefs) > 0 {
		return OutcomeHealthCompleted
	}
	return OutcomeHealthHealthy
}

func OutcomeHealthForTeamWork(item TeamWorkItem) OutcomeHealthState {
	if item.State == TeamWorkStateArchived {
		return OutcomeHealthArchived
	}
	if item.NeedsOperator {
		return OutcomeHealthBlocked
	}
	switch item.State {
	case TeamWorkStateOutputReady:
		return OutcomeHealthCompleted
	case TeamWorkStateNeedsOperator:
		return OutcomeHealthBlocked
	case TeamWorkStateDegraded:
		return OutcomeHealthDegraded
	case TeamWorkStateRunning, TeamWorkStateReviewing:
		return OutcomeHealthRunning
	case TeamWorkStateNew, TeamWorkStateBriefed, TeamWorkStateQueued, TeamWorkStatePaused:
		return OutcomeHealthWaiting
	default:
		return OutcomeHealthHealthy
	}
}

func OutcomeHealthForTeamStatusEvent(item TeamStatusEvent) OutcomeHealthState {
	if item.State == TeamWorkStateArchived {
		return OutcomeHealthArchived
	}
	if len(item.BlockedBy) > 0 {
		return OutcomeHealthBlocked
	}
	switch item.State {
	case TeamWorkStateOutputReady:
		return OutcomeHealthCompleted
	case TeamWorkStateNeedsOperator:
		return OutcomeHealthBlocked
	case TeamWorkStateDegraded:
		return OutcomeHealthDegraded
	case TeamWorkStateRunning, TeamWorkStateReviewing:
		return OutcomeHealthRunning
	case TeamWorkStateNew, TeamWorkStateBriefed, TeamWorkStateQueued, TeamWorkStatePaused:
		return OutcomeHealthWaiting
	default:
		return OutcomeHealthHealthy
	}
}

func OutcomeHealthForRunStatus(status string) OutcomeHealthState {
	switch strings.ToLower(strings.TrimSpace(status)) {
	case "archived":
		return OutcomeHealthArchived
	case "pending", "queued", "new", "briefed", "paused":
		return OutcomeHealthWaiting
	case "running", "reviewing":
		return OutcomeHealthRunning
	case "completed", "succeeded", "success", "output_ready":
		return OutcomeHealthCompleted
	case "degraded", "needs_attention":
		return OutcomeHealthDegraded
	case "failed", "blocked", "needs_operator", "cancelled", "canceled":
		return OutcomeHealthBlocked
	default:
		return OutcomeHealthHealthy
	}
}

// AggregateOutcomeHealth returns the most actionable state across an Outcome.
// Archived only wins when every supplied state is archived.
func AggregateOutcomeHealth(states ...OutcomeHealthState) OutcomeHealthState {
	if len(states) == 0 {
		return OutcomeHealthHealthy
	}
	allArchived := true
	seen := make(map[OutcomeHealthState]bool, len(states))
	for _, state := range states {
		seen[state] = true
		if state != OutcomeHealthArchived {
			allArchived = false
		}
	}
	if allArchived {
		return OutcomeHealthArchived
	}
	for _, state := range []OutcomeHealthState{
		OutcomeHealthBlocked,
		OutcomeHealthDegraded,
		OutcomeHealthRunning,
		OutcomeHealthCompleted,
		OutcomeHealthWaiting,
		OutcomeHealthHealthy,
	} {
		if seen[state] {
			return state
		}
	}
	return OutcomeHealthHealthy
}

File: interface/lib/outcomeHealth.ts
export type OutcomeHealthState =
  | "healthy"
  | "waiting"
  | "running"
  | "degraded"
  | "blocked"
  | "completed"
  | "archived";

export function normalizeOutcomeHealth(value: unknown): OutcomeHealthState {
  const raw = typeof value === "string" ? value.trim().toLowerCase() : "";
  if (
    raw === "healthy" ||
    raw === "waiting" ||
    raw === "running" ||
    raw === "degraded" ||
    raw === "blocked" ||
    raw === "completed" ||
    raw === "archived"
  ) {
    return raw;
  }
  return "healthy";
}

export function outcomeHealthLabel(health: OutcomeHealthState) {
  return health.charAt(0).toUpperCase() + health.slice(1);
}

export function outcomeHealthFromRunStatus(value: unknown): OutcomeHealthState {
  const raw = typeof value === "string" ? value.trim().toLowerCase() : "";
  if (raw === "archived") return "archived";
  if (["pending", "queued", "new", "briefed", "paused"].includes(raw)) return "waiting";
  if (["running", "reviewing"].includes(raw)) return "running";
  if (["completed", "succeeded", "success", "output_ready"].includes(raw)) return "completed";
  if (["degraded", "needs_attention"].includes(raw)) return "degraded";
  if (["failed", "blocked", "needs_operator", "cancelled", "canceled"].includes(raw)) return "blocked";
  return "healthy";
}

export function aggregateOutcomeHealth(states: OutcomeHealthState[]): OutcomeHealthState {
  if (states.length === 0) return "healthy";
  if (states.every((state) => state === "archived")) return "archived";
  for (const state of ["blocked", "degraded", "running", "completed", "waiting", "healthy"] as const) {
    if (states.includes(state)) return state;
  }
  return "healthy";
}

export function outcomeHealthClassName(health: OutcomeHealthState) {
  if (health === "blocked") return "border-cortex-danger/35 bg-cortex-danger/10 text-cortex-danger";
  if (health === "degraded" || health === "waiting") {
    return "border-cortex-warning/40 bg-cortex-warning/10 text-cortex-warning";
  }
  if (health === "running" || health === "completed") {
    return "border-cortex-primary/35 bg-cortex-primary/10 text-cortex-primary";
  }
  if (health === "archived") return "border-cortex-border bg-cortex-bg text-cortex-text-muted";
  return "border-cortex-success/35 bg-cortex-success/10 text-cortex-success";
}
